File: support/automail.php
<?
/*
	osTicket, Open Source Support Ticket System
	http://www.osticket.com

	Copyright (C) 2003 osTicket, Joseph Shain

	Released under the GNU General Public License
*/

include_once("class.ticket.php");
include_once("config.php");

if (!function_exists('imap_open')) {
    $err = "IMAP functions are not available.<br>";
}
else {

function get_mime_type(&$structure) {
	$primary_mime_type = array("TEXT", "MULTIPART", "MESSAGE", "APPLICATION", "AUDIO", "IMAGE", "VIDEO", "OTHER");
	
	if($structure->subtype) {
		return $primary_mime_type[(int) $structure->type] . '/' . $structure->subtype;
	}
	return "TEXT/PLAIN";
}

function get_part($stream, $msg_number, $mime_type, $structure = false, $part_number = false) {
	if(!$structure) {
		$structure = imap_fetchstructure($stream, $msg_number);
	}
	if($structure) {
		if($mime_type == get_mime_type($structure)) {
			if(!$part_number) {
				$part_number = "1";
			}
			$text = imap_fetchbody($stream, $msg_number, $part_number);
			if($structure->encoding == 3) {
				return imap_base64($text);
			} else if($structure->encoding == 4) {
				return imap_qprint($text);
			} else {
				return $text;
			}
		}
		if($structure->type == 1) /* multipart */ {
			while(list($index, $sub_structure) = each($structure->parts)) {
				if($part_number) {
					$prefix = $part_number . '.';
				}
				$data = get_part($stream, $msg_number, $mime_type, $sub_structure, $prefix . ($index + 1));
				if($data) {
					return $data;
				}
			}
		}
	}
	return false;
}

$cat_res = mysql_query("SELECT * FROM ticket_categories");
while ($cat_row = mysql_fetch_array($cat_res)) {
if ($cat_row[pophost]) {
    $mbox = @imap_open("{" . $cat_row[pophost] . "/pop3:110}INBOX", $cat_row[popuser], $cat_row[poppass]);
    if (!$mbox) {
    	$mbox = @imap_open("{" . $cat_row[pophost] . "/pop3/notls}INBOX", $cat_row[popuser], $cat_row[poppass]);
    }
    	
    if (!$mbox) {
	    $err .= "Unable to open mailbox for category \"$cat_row[name]\".<br>";
	}
	else {
	$curmsg = 1;

	while ($curmsg <= @imap_num_msg($mbox)) {
		$body = get_part($mbox, $curmsg, "TEXT/PLAIN");
		if (!$body) {
			$body = get_part($mbox, $curmsg, "TEXT/HTML");
		}
		if (!$body) {
			continue;
		}
		$head = imap_headerinfo($mbox, $curmsg, 800, 800);
		$email = $head->reply_toaddress;
		
		if (strpos($email, "<")) {
			$email = eregi_replace(".*<(.*)>.*", "\\1", $email);
        }
			
        $name = $head->fromaddress;
///ADDED by KVAS: character fix
$decoded1 = imap_mime_header_decode($name);
$name = $decoded1[0]->text;
////END
		if (strpos($name, "<")) {
			$name = eregi_replace("(.*) <.*>.*", "\\1", $name);
        }
        if (!$name) {
            $name = $email;
        }
            			
		$subject = $head->fetchsubject;
///ADDED by KVAS: character fix
    $decoded = imap_mime_header_decode($subject);
    $subject = $decoded[0]->text;
    $head->fetchsubject = $subject;
/////END	
        $subject = !$subject ? "[No Subject]": $subject;

        $eml_headers = imap_fetchheader($mbox, $curmsg, FT_PREFETCHTEXT);
        $x_pri = split("\n", $eml_headers);
        foreach ($x_pri as $item) {
            $arr = split(": ", $item);
            if (eregi("x-priority", $arr[0])) {
                if (strstr($arr[1], "1") or strstr($arr[1], "2")) {
                    $pri = 3;
                }
                elseif (strstr($arr[1], "4") or strstr($arr[1], "5")) {
                    $pri = 1;
                }
            }
            elseif (eregi("X-Originating-IP", $arr[0])) {
                $ip = $arr[1];
                $ip = eregi_replace("\[(.*)\]", "\\1", $ip);
            }
        }
        if (!$pri) {
            $pri = 2;
        }
            
        $body = ltrim(rtrim($body));
        
        if ($config[remove_original] and $config[remove_tag] and strpos($body, $config[remove_tag])) {
            $lines = split($config[remove_tag], $body);
            $body = $lines[0];
        }

        $c = mysql_fetch_array(mysql_query("SELECT COUNT(*) AS cnt FROM tickets WHERE email='$email' AND status='open'"));
        if ($c[cnt] > $config[ticket_max]) {
            $config[limit_msg] = str_replace("%local_email", $cat_row[email], $config[limit_msg]);
            $config[limit_msg] = str_replace("%ticket_max", $config[ticket_max], $config[limit_msg]);
            
            if ($config[limit_response]) {
                mail($email, $config[limit_subj], $config[limit_msg], "From: $config[limit_email]\n");
            }
            die();
        }
        
    	if (!ereg ("[[][#][0-9]{6}[]]", $head->fetchsubject)) {
            $ticket_id = CreateTicket($subject, $name, $email, $cat_row[ID], "", $pri, $ip);
    	}
        else {
    		$ticket_id = substr(strstr($head->fetchsubject, "[#"), 2, 6);
    	}
    	
    	$iid = PostMessage($ticket_id, $body, $eml_headers, false);
    	
    	if (is_dir($config[attachment_dir]) and $config[accept_attachments]) {
    	$struct = imap_fetchstructure($mbox, $curmsg);
        for ($i = 0; $q < count($struct->parts); ++$q) {
            $filename="";
            $section = $struct->parts[$q];

            $param = $section->dparameters;
            for ($x = 0; $x <= count($param); ++$x) {
                if (eregi("filename", $param[$x]->attribute)) {
                    $filename = $param[$x]->value;
///ADDED by KVAS: character fix
	$decoded_filename = imap_mime_header_decode($param[$x]->value);
	$filename = $decoded_filename[0]->text;
///END
                }
            }
            $part = imap_fetchbody($mbox, $curmsg, $q+1);
				$ext = preg_replace("/.*\.(.{3,4})$/", "$1",$filename);
            if ($filename && stristr($config[filetypes],".$ext;")) {
                if ($section->encoding == "3") {
                    $part = base64_decode($part);
                }
                if (!$config[attachment_size] or ($section->bytes <= $config[attachment_size])) {
							
						  mt_srand(time());
               	  $rand=mt_rand(100000, 999999); //six chars.
                    $file = fopen("$config[attachment_dir]/$rand"."_$filename", "w+");
                    fputs($file, $part);
                    fclose($file);
                    mysql_query("INSERT INTO ticket_attachments (ticket, ref, filename, type) VALUES ($ticket_id, $iid, '$rand" . "_$filename', 'q')");
                }
            }
        }
        }
        
    	imap_delete($mbox, $curmsg);
		$curmsg++;
	}
	@imap_expunge($mbox);
	@imap_close($mbox);
	}
	}
}
}
?>

File: support/automail.pl
#!/usr/local/bin/perl

#######################################################
#
#	osTicket, Open Source Support Ticket System
#	http://www.osticket.com
#
#	Copyright (C) 2003 osTicket, Joseph Shain
#
#	Released under the GNU General Public License
#
#######################################################
#
#	Setup automail.pl Email Gateway for aliases
#
#	Scroll down to configure your path to
#	Sendmail and MySQL settings.
#
#######################################################

sub is_valid {
    $ticket = $_[0];

    $sth = $dbh->prepare("SELECT * FROM tickets WHERE ID=$ticket");
    $sth->execute();

    $hash = $sth->fetchrow_hashref;
    $sth->finish();

    return $hash->{ID};
}

sub sendmail {
    ($to, $from, $subj, $body) = @_;
    open(MAIL, "|$config{'mailprog'} -t");
    print MAIL "To: $to\n";
    print MAIL "From: $from\n";
    print MAIL "Subject: $subj\n\n";
    print MAIL "$body\n";
    close(MAIL);
}

sub keyword {
    $_[0] =~ s/%([_|a-zA-Z][a-zA-Z|\d|_]*)/${$1}/gi;
}

#configuration
%config = (mailprog => '/usr/lib/sendmail',
           database_host => 'localhost',
           database_name => 'Your_Database__Here',
           database_user => 'Your_Username_Here',
           database_pass => 'Your_Password_Here');

use DBI;
$dbh = DBI->connect("DBI:mysql:$config{'database_name'}:$config{'database_host'}", $config{'database_user'}, $config{'database_pass'});

#database configuration
$sth = $dbh->prepare("SELECT * FROM ticket_config");
$sth->execute();
$myconfig = $sth->fetchrow_hashref;

$url = $myconfig->{root_url};

#email message
while (<STDIN>) {
    push @message, $_;
    $eml .= $_;
}

#parse headers
foreach (@message) {
    push @headers, $_;
    last if (/^\s$/ || /^$/);
    if (/oundary=/) {
        $attachment_info = $_;
        $attachment = 1;
    }
    else {
        $attachment = 0;
    }
    $_ =~ s/:\s/:/g;
    if (/:/) {
        @vars = split(':', $_, 2);
        if ($vars[1]) {
            chop($header{$vars[0]} = $vars[1]);
        }
    }
}

$eml_subj = $header{'Subject'};
$user_email = $header{'From'};
$user_email =~ s/"//g;
$local_email = $header{'To'};

$eml_subj =~ s/\s*Re:\s*//gi if ($eml_subj);
$eml_subj = "[No Subject]" if (!$eml_subj);
$user_email =~ s/"/\"/g;

@ticket = split('\[#', $eml_subj, 2);
chop($ticket[1]);
$ticket = int($ticket[1]);

#is address banned?
$sth = $dbh->prepare("SELECT * FROM ticket_banlist");
$sth->execute();
while ($ban = $sth->fetchrow_hashref) {
    $banlist = $ban->{value};
    if ($user_email =~ /$banlist/i) {
        die();
    }
}

#format email priority to osticket priority.
$pri = $header{'X-Priority'};
if ($pri == 1 or $pri == 2) {
    $pri = 3;
}
elsif ($pri == 3) {
    $pri = 2;
}
elsif ($pri == 4 or $pri == 5) {
    $pri = 1;
}
else {
    $pri = 2;
}

#attachment info
($i, $boundary) = split("\"", $attachment_info);

#get filenames and base64 codes for file attachments
if ($boundary) {
    @eml = split($boundary, $eml);
    for ($i = 0; $i <= $#eml; ++$i) {
        if ($eml[$i] =~ /filename/) {
            @lines = split("\n", $eml[$i]);

            $encoding = $eml[$i];
            $encoding =~ s/Content-Transfer-Encoding:\s(.*)\n/$1/i;
            push @encoding, $encoding;

            $feed = 0;
            $code = "";
            foreach (@lines) {
                if ($feed) {
                    $code .= "$_\n";
                }

                if ($_ =~ /filename/) {
                    $_ =~ s/.*filename="(.*)"/$1/;
                    push @files, $_;
                    $feed = 1;
                }
            }
            push @codes, $code;
        }
        else {
            push @bodies, $eml[$i];
        }
    }
}
else {
    for ($i = $#headers + 1; $i <= $#message; $i++) {
        $body .= $message[$i];
    }
}

#find body (if multi-part email)
foreach(@bodies) {
    @lines = split("\n", $_);
    foreach (@lines) {
        if (/oundary=/) {
            $b = $_;
        }
    }
    ($i, $boundary) = split("\"", $b);
    if ($boundary) {
        @ibodies = split($boundary, $_);
    }
    else {
        $ibodies[0] = $_;
    }

    foreach(@ibodies) {
        if ($#bodies == 0) {
            $type = ".*";
        }
        else {
            $type = "plain";
        }
        if (/content-type:\stext\/$type/i) {
            $body = $_;
            @lines = split("\n", $body);
            $body = "";
            $f = -1;
            foreach(@lines) {
                if ($f) {
                    $body .= "$_\n";
                }
                if (!$_) {
                    ++$f;
                }
            }

        }
    }
}

#delete original message
if ($myconfig->{remove_original} && $myconfig->{remove_tag} && index($body, $myconfig->{remove_tag})) {
    $remove_tag = $myconfig->{remove_tag};
    @lines = split(/$remove_tag/i, $body);
    $body = $lines[0];

    @lines = split(/-*[\n|\s|\0]Original Message[\n|\s|\0]-*/, $body);
    $body = $lines[0];
}

#right and left trim
$body =~ s/<script[^>]*?>.*?<\/script>//g;
$body =~ s/<[\/!]*?[^<>]*?>//g;
$body =~ s/'/\\'/g;
$body =~ s/--$//g;
$body =~ s/[\n|<br>]*$//gi;
$body =~ s/^(.*)>/<$1>/i;
$body =~ s/^[\n|<br>]*//gi;

#get category
$sth = $dbh->prepare("SELECT * FROM ticket_categories");
$sth->execute();
while ($row_hash = $sth->fetchrow_hashref) {
    if ($local_email =~ $row_hash->{email}) {
        $local_email = $row_hash->{email};
        $cat = $row_hash->{ID};
        $signature = $row_hash->{signature};
    }
}

#add signature/remove tag
if ($signature) {
    $signature = "\n\n$signature";
}
if ($myconfig->{remove_original}) {
    $remove_tag = $myconfig->{remove_tag} . "\n\n";
}

#save email headers
if ($myconfig->{save_headers}) {
    $eml_headers = $eml;
}
$eml_headers = $dbh->quote($eml_headers);

#get greenwich mean time
use POSIX;
@gm = gmtime();
$gmtime = POSIX::mktime($gm[0], $gm[1], $gm[2], $gm[3], $gm[4], $gm[5]);

@oldgm = gmtime(time()-75);
$gmoldtime = $oldgm[5] . $oldgm[4] . $oldgm[3] . $oldgm[2] . $oldgm[1] . $oldgm[0];

@what = split('<', $user_email, 2);
$email = $what[1];
$email =~ s/\>//g;
$email =~ s/\'//g;

$name = $what[0];
$name =~ s/\s$//;
if (!$name) {
     $name = $email;
}
elsif (!$email) {
     $email = $name;
}

# flood protection, find all posts made by this email in the last 10 minutes
@smoog_what = split('<', $user_email, 2);
$smoog_email = $smoog_what[1];
$smoog_email =~ s/\>//g;
$smoog_email =~ s/\'//g;

$smoog_mysqltime = $gmtime - 75;
$smoog_earlieststamp = -1;
$smoog_messagesfound = 0;
$smoog_query = "SELECT * FROM tickets WHERE email='$smoog_email' ORDER BY timestamp ASC";
$smoog_tickets = $dbh->prepare($smoog_query);
$smoog_tickets->execute();

while ($smoog_t = $smoog_tickets->fetchrow_hashref) {
	  $smoog_query = "SELECT UNIX_TIMESTAMP(timestamp) as unitimestamp FROM ticket_messages WHERE ticket='" . $smoog_t->{ID} . "' AND timestamp > '$gmoldtime'";
      $smoog_messages = $dbh->prepare($smoog_query);
      $smoog_messages->execute();
      while ($m = $smoog_messages->fetchrow_hashref) {
      		if ($smoog_earlieststamp == -1 || $smoog_earlieststamp > $m->{unitimestamp}) {
      			$smoog_earlieststamp = $m->{unitimestamp};
      		} 
			$smoog_messagesfound++;
     }
}
# done figuring out recent message count, now lets check the rate.
$smoog_msgrate = (60 * $smoog_messagesfound) / ($gmtime - $smoog_earlieststamp); 
if (!length($smoog_email) || $smoog_msgrate > 5) { # you can change the number here, its the max num of messages per min sent by 1 email.
	die();
}
# print 'Message rate: ' . $smoog_msgrate . ', messages found: ' . $smoog_messagesfound . ', query: '. $smoog_query . "\n, email: $smoog_email $gmtime $user_email";
# die();
# end flood protection

#create new ticket
if (!$ticket) {
    do {
        $min = 100000;
        $max = 999999;
        $ticket = int(rand($max - $min + 1)) + $min;
    } while (is_valid($ticket));

    chdir($myconfig->{attachment_dir});

    @what = split('<', $user_email, 2);
    $email = $what[1];
    $email =~ s/\>//g;
    $email =~ s/\'//g;

    $name = $what[0];
    $name =~ s/\s$//;
    if (!$name) {
        $name = $email;
    }
    elsif (!$email) {
        $email = $name;
    }

    $sth = $dbh->prepare("SELECT COUNT(*) AS tq FROM tickets WHERE email='$email' AND status='open'");
    $sth->execute();
    $row_hash = $sth->fetchrow_hashref;
    $tq = $row_hash->{tq};

    if ($tq < $myconfig->{ticket_max}) {
        $eml_subj = $dbh->quote($eml_subj);
        $name = $dbh->quote($name);

        $sth = $dbh->prepare("SELECT UNIX_TIMESTAMP(timestamp) AS timestamp FROM tickets WHERE email='$email' ORDER BY timestamp DESC");
        $sth->execute();
        $row_hash = $sth->fetchrow_hashref;
        $interval = time() - $row_hash->{timestamp};
        $ip =~ s/\[(.*)\]/$1/;

        $sth = $dbh->prepare("INSERT INTO tickets (subject, name, email, cat, status, ID, priority, ip, timestamp) VALUES ($eml_subj, $name, '$email', '$cat', 'open', $ticket, $pri, '" . $header{'X-Originating-IP'} . "', FROM_UNIXTIME('$gmtime') + 0)");
        $sth->execute();

        $sth = $dbh->prepare("INSERT INTO ticket_messages (ticket, message, headers, timestamp) VALUES ($ticket, '$body', $eml_headers, FROM_UNIXTIME('$gmtime') + 0)");
        $sth->execute();

        if ($interval > $myconfig->{min_interval}) {
            keyword($myconfig->{ticket_subj});
            keyword($myconfig->{ticket_msg});
            if ($myconfig->{ticket_response}) {
                sendmail($user_email, $local_email, $myconfig->{ticket_subj}, $remove_tag . $myconfig->{ticket_msg} . $signature);
            }
        }
    }
    #if over ticket limit
    else {
        $ticket_max = $myconfig->{ticket_max};

        keyword($myconfig->{limit_subj});
        keyword($myconfig->{limit_msg});
        if ($myconfig->{limit_response}) {
            sendmail($user_email, $myconfig->{limit_email}, $myconfig->{limit_subj}, $remove_tag . $myconfig->{limit_msg} . $signature);
        }
    }
}
#post message
else {
    $sth = $dbh->prepare("UPDATE tickets SET status = 'open' WHERE ID=$ticket");
    $sth->execute();

    $sth = $dbh->prepare("INSERT INTO ticket_messages (ticket, message, headers, timestamp) VALUES ($ticket, '$body', $eml_headers, FROM_UNIXTIME('$gmtime') + 0)");
    $sth->execute();

    keyword($myconfig->{message_subj});
    keyword($myconfig->{message_msg});

    if ($myconfig->{message_response}) {
        sendmail($user_email, $local_email, $myconfig->{message_subj}, $remove_tag . $myconfig->{message_msg} . $signature);
    }
}

#alert owner
if ($myconfig->{alert_new}) {
    keyword($myconfig->{alert_subj});

    $email = $user_email;
    $content = $body;

    keyword($myconfig->{alert_msg});
    sendmail($myconfig->{alert_user}, $myconfig->{alert_email}, $myconfig->{alert_subj}, $myconfig->{alert_msg});
}

#upload attachments
if ($myconfig->{accept_attachments} && $myconfig->{attachment_dir}) {
    use MIME::Base64;
    chdir($myconfig->{attachment_dir});

    $sth = $dbh->prepare("SELECT LAST_INSERT_ID() AS iid");
    $sth->execute();
    $row_hash = $sth->fetchrow_hashref;
    $iid = $row_hash->{iid};

    for ($i = 0; $i <= $#files; ++$i) {
        $id= int(rand(999999-100000+1))+100000;
        $filename=$id."_$files[$i]";
        open(ATTACH, ">$filename");
        binmode(ATTACH);
        $codes[$i] =~ s/Content-Transfer-Encoding:\s(.*)\n//i;

        if ($encoding[$i] =~ /base64/i) {
            $codes[$i] = decode_base64($codes[$i])
        }

        if (length($codes[$i]) < $myconfig->{attachment_size}) {
            print ATTACH $codes[$i];
        }
        close(ATTACH);

        $sth = $dbh->prepare("INSERT INTO ticket_attachments (ticket, ref, filename, type) VALUES ($ticket, $iid, '$filename', 'q')");
        $sth->execute();
    }
}

$sth->finish();
$dbh->disconnect();
